pkg/controller/infinispan/resources/resource_controller.go

```diff
@@ -0,0 +1,157 @@
+package resources
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/go-logr/logr"
+	v1 "github.com/infinispan/infinispan-operator/pkg/apis/infinispan/v1"
+	consts "github.com/infinispan/infinispan-operator/pkg/controller/constants"
+	"github.com/infinispan/infinispan-operator/pkg/controller/infinispan"
+	kube "github.com/infinispan/infinispan-operator/pkg/kubernetes"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+//
+type Resource interface {
+	Process() (reconcile.Result, error)
+}
+
+type ReconcileType struct {
+	ObjectType            runtime.Object
+	GroupVersion          schema.GroupVersion
+	GroupVersionSupported bool
+}
+
+func (rt ReconcileType) GroupVersionKind() schema.GroupVersionKind {
+	return schema.GroupVersionKind{
+		Group:   rt.GroupVersion.Group,
+		Version: rt.GroupVersion.Version,
+		Kind:    rt.Kind(),
+	}
+}
+
+func (rt ReconcileType) Kind() string {
+	return reflect.TypeOf(rt.ObjectType).Elem().Name()
+}
+
+type Reconciler interface {
+	// The k8 structs being handled by this controller
+	Types() []*ReconcileType
+	// Events for the struct handled by this controller
+	EventsPredicate() predicate.Predicate
+	// Create a new instance of Infinispan wrapping the actual k8s type
+	InfinispanInstance(infinispan *v1.Infinispan, ctrl *Controller, kube *kube.Kubernetes, log logr.Logger) Resource
```

Comment:
How about `ResourceInstance` as we're returning an implementation of the `Resource` interface?

Reply:
Yes, agreed. Will change